# –ú–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏–µ

# **–í–≤–µ–¥–µ–Ω–∏–µ**

- –ü–æ–ª–µ–∑–Ω—ã–µ –º–∞—Ç–µ—Ä–∏–∞–ª—ã
    
    [10.5 Extracting interesting regions: Differential methylation and segmentation | Computational Genomics with R](https://compgenomr.github.io/book/extracting-interesting-regions-differential-methylation-and-segmentation.html)
    
    [DNA methylation data by sequencing: experimental approaches and recommendations for tools and pipelines for data analysis - Clinical Epigenetics](https://clinicalepigeneticsjournal.biomedcentral.com/articles/10.1186/s13148-019-0795-x)
    
    [Running Bismark - Bismark](https://felixkrueger.github.io/Bismark/bismark/)
    


–¶–µ–ª—å—é –¥–∞–Ω–Ω–æ–≥–æ –¥–æ–º–∞—à–Ω–µ–≥–æ –∑–∞–¥–∞–Ω–∏—è —è–≤–ª—è–µ—Ç—Å—è –∏–∑—É—á–µ–Ω–∏–µ –≥–ª–æ–±–∞–ª—å–Ω–æ–≥–æ –∏–∑–º–µ–Ω–µ–Ω–∏—è —É—Ä–æ–≤–Ω—è CpG –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è –î–ù–ö –ø—Ä–∏ —Ä–∞–Ω–Ω–µ–º —ç–º–±—Ä–∏–æ–Ω–∞–ª—å–Ω–æ–º —Ä–∞–∑–≤–∏—Ç–∏–∏ –º—ã—à–∏. –°—á–∏—Ç–∞–µ—Ç—Å—è, —á—Ç–æ –ø—Ä–∏ —Ä–∞–∑–≤–∏—Ç–∏–∏ —ç–º–±—Ä–∏–æ–Ω–∞ –ø—Ä–æ–∏—Å—Ö–æ–¥—è—Ç —Ç–∞–∫ –Ω–∞–∑—ã–≤–∞–µ–º—ã–µ –≤–æ–ª–Ω—ã –¥–µ–º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è-–º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è, —Ç.–µ. –Ω–∞ —Ä–∞–Ω–Ω–∏—Ö —Å—Ç–∞–¥–∏—è—Ö CpG –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏–µ —É–º–µ–Ω—å—à–∞–µ—Ç—Å—è –¥–æ –Ω–µ–∫–æ—Ç–æ—Ä–æ–≥–æ –º–∏–Ω–∏–º—É–º–∞ (–æ–∫–æ–ª–æ 25%), –∞ –∑–∞—Ç–µ–º –ø–æ –º–µ—Ä–µ –¥–∏—Ñ—Ñ–µ—Ä–µ–Ω—Ü–∏—Ä–æ–≤–∫–∏ —Ç–∫–∞–Ω–µ–π –æ–Ω–æ —Å–∏–ª—å–Ω–æ —É–≤–µ–ª–∏—á–∏–≤–∞–µ—Ç—Å—è (–æ–∫–æ–ª–æ 90%) –∏ –æ—Å—Ç–∞–µ—Ç—Å—è —Ç–∞–∫–∏–º –Ω–∞ –ø—Ä–æ—Ç—è–∂–µ–Ω–∏–∏ –≤—Å–µ–π –∂–∏–∑–Ω–∏ –æ—Ä–≥–∞–Ω–∏–∑–º–∞:

![Picture1.png](figures/Picture1.png)

–î–ª—è –≤—ã–ø–æ–ª–Ω–µ–Ω–∏—è –¥–∞–Ω–Ω–æ–π –∑–∞–¥–∞—á–∏ –º—ã –±—É–¥–µ–º –∏–∑—É—á–∞—Ç—å —Å–ª–µ–¥—É—é—â–∏–µ –æ–±—Ä–∞–∑—Ü—ã WGBS (Whole genome bisulfite sequencing), —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—â–∏–µ —Ä–∞–∑–Ω—ã–º —Å—Ç–∞–¥–∏—è–º —ç–º–±—Ä–∏–æ–Ω–∞–ª—å–Ω–æ–≥–æ —Ä–∞–∑–≤–∏—Ç–∏—è –º—ã—à–∏:

- 8cell_rep1_WGBS ‚Äì 8-–∫–ª–µ—Ç–æ—á–Ω—ã–π —ç–º–±—Ä–∏–æ–Ω, –ø—Ä–∏–º–µ—Ä–Ω–æ 2.25 –¥–Ω—è –ø–æ—Å–ª–µ –æ–ø–ª–æ–¥–æ—Ç–≤–æ—Ä–µ–Ω–∏—è —è–π—Ü–µ–∫–ª–µ—Ç–∫–∏
- ICM_rep1_WGBS ‚Äì –≤–Ω—É—Ç—Ä–µ–Ω–Ω—è—è –∫–ª–µ—Ç–æ—á–Ω–∞—è –º–∞—Å—Å–∞ –±–ª–∞—Å—Ç–æ—Ü–∏—Å—Ç–∞, –ø—Ä–∏–º–µ—Ä–Ω–æ 3.5 –¥–Ω—è –ø–æ—Å–ª–µ –æ–ø–ª–æ–¥–æ—Ç–≤–æ—Ä–µ–Ω–∏—è —è–π—Ü–µ–∫–ª–µ—Ç–∫–∏
- Epiblast_rep1_WGBS ‚Äì —Å—Ç–∞–¥–∏—è —ç–ø–∏–±–ª–∞—Å—Ç–∞, –ø—Ä–∏–º–µ—Ä–Ω–æ 6.5 –¥–Ω–µ–π –ø–æ—Å–ª–µ –æ–ø–ª–æ–¥–æ—Ç–≤–æ—Ä–µ–Ω–∏—è —è–π—Ü–µ–∫–ª–µ—Ç–∫–∏

–í —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—â–µ–π —Å—Ç–∞—Ç—å–µ [PMID: 28959968] –∞–≤—Ç–æ—Ä—ã –ø—Ä–∏–≤–æ–¥—è—Ç —Å–ª–µ–¥—É—é—â–∏–µ —Ä–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏—è –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è —Ü–∏—Ç–æ–∑–∏–Ω–æ–≤:

![Picture2.png](figures/Picture2.png)

# –ü–æ–¥–≥–æ—Ç–æ–≤–∏–º —Ä–∞–±–æ—á–µ–µ –ø—Ä–æ—Å—Ç—Ä–∞–Ω—Å—Ç–≤–æ

–ü–µ—Ä–µ–¥–∞ –Ω–∞—á–∞–ª–æ–º –ø–æ–¥–≥–æ—Ç–æ–≤–∏–º —Ä–∞–±–æ—á–µ–µ –ø—Ä–æ—Å—Ç—Ä–∞–Ω—Å—Ç–≤–æ, –∞ –∏–º–µ–Ω–Ω–æ —Å–æ–∑–¥–∞–¥–∏–º —Ä–∞–±–æ—á—É—é –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é –∏ –≤—Å–µ –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –≤–ª–æ–∂–µ–Ω–Ω—ã–µ —Ñ–∞–π–ª—ã.

```bash
# –°–æ–∑–¥–∞–¥–∏–º —Ä–∞–±–æ—á—É—é –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏—é –∏ –ø–µ—Ä–µ–º–µ—Å—Ç–∏–º—Å—è –≤ –Ω–µ—ë
$ mkdir -p omics_hw/methylation
$ cd omics_hw/methylation

# C–æ–∑–¥–∞–¥–∏–º –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏–∏
$ mkdir fastq_files # –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏—è —Å —á–∏—Å—Ç—ã–º–∏ .fastq —Ñ–∞–π–ª—ã
$ mkdir scripts # –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏—è —Å–æ —Å–∫—Ä–∏–ø—Ç–∞–º–∏ —Å –ø–æ–º–æ—â—å—é –∫–æ—Ç–æ—Ä—ã—Ö –º—ã –±—É–¥–µ–º –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞—Ç—å –¥–∞–Ω–Ω—ã–µ
$ mkdir references # –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏—è c –≥–µ–Ω–æ–º–Ω—ã–º —Ä–µ—Ñ–µ—Ä–µ–Ω—Å–æ–º, –∞–Ω–Ω–æ—Ç–∞—Ü–∏–µ–π –∏ –∏–Ω–¥–µ–∫—Å–∞–º–∏ –¥–ª—è –≥–µ–Ω–æ–º–∞
$ mkdir fastq_trimmed # –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏—è —Å —Ç—Ä–∏–º–º–∏—Ä–æ–≤–∞–Ω—ã–º–∏ .fastq —Ñ–∞–π–ª–∞–º–∏
$ mkdir qc # –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏—è —Å –æ—Ç—á—ë—Ç–∞–º–∏ –æ –∫–∞—á–µ—Å—Ç–≤–µ –Ω–∞—à–∏—Ö –¥–∞–Ω–Ω—ã—Ö
```

# –°–∫–∞—á–∏–≤–∞–Ω–∏–µ —Ñ–∞–π–ª–æ–≤

–°–∫–∞—á–∞–µ–º –¥–∞–Ω–Ω—ã–µ **SRP078328 ****–ø—Ä–∏ –ø–æ–º–æ—â–∏ [SRA Explorer](https://sra-explorer.info/#). –î–ª—è —ç—Ç–æ–≥–æ —Å–æ–∑–¥–∞–¥–∏–º —Å–∫—Ä–∏–ø—Ç `load_fastq.sh` –≤ –¥–∏—Ä—Ä–µ–∫—Ç–æ—Ä–∏–∏ `scripts`. –î–æ–±–∞–≤–∏–º `$1` –ø–µ—Ä–µ–¥ –Ω–∞–∑–≤–∞–Ω–∏–µ–º —Ñ–∞–π–ª–∞ —á—Ç–æ–±—ã —É–∫–∞–∑–∞—Ç—å –Ω–µ–æ–±—Ö–æ–¥–∏–º—É—é –Ω–∞–º –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é. –ê —Ç–∞–∫ –∂–µ —É–¥–∞–ª–∏–º –∫–æ–Ω–∫–∞—Ç–µ–Ω–∏—Ä–æ–≤–∞–Ω—ã–µ —Ñ–∞–π–ª—ã, –∫–æ—Ç–æ—Ä—ã–µ –ø–æ –∫–∞–∫–æ–π-—Ç–æ –ø—Ä–∏—á–∏–Ω–µ –≤—ã–¥–∞—ë—Ç –Ω–∞–º SRA Explorer

```bash
#!/usr/bin/env bash
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/007/SRR5836477/SRR5836477_1.fastq.gz -o $1/SRR5836477_GSM2229984_TE_rep1_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/007/SRR5836477/SRR5836477_2.fastq.gz -o $1/SRR5836477_GSM2229984_TE_rep1_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/006/SRR5836476/SRR5836476_1.fastq.gz -o $1/SRR5836476_GSM2229983_ICM_rep2_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/006/SRR5836476/SRR5836476_2.fastq.gz -o $1/SRR5836476_GSM2229983_ICM_rep2_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/004/SRR5836474/SRR5836474_1.fastq.gz -o $1/SRR5836474_GSM2229981_8cell_rep2_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/004/SRR5836474/SRR5836474_2.fastq.gz -o $1/SRR5836474_GSM2229981_8cell_rep2_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/005/SRR5836475/SRR5836475_1.fastq.gz -o $1/SRR5836475_GSM2229982_ICM_rep1_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/005/SRR5836475/SRR5836475_2.fastq.gz -o $1/SRR5836475_GSM2229982_ICM_rep1_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR382/002/SRR3824222/SRR3824222_1.fastq.gz -o $1/SRR3824222_GSM2229986_Epiblast_rep1_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR382/002/SRR3824222/SRR3824222_2.fastq.gz -o $1/SRR3824222_GSM2229986_Epiblast_rep1_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_1.fastq.gz -o $1/SRR5836473_GSM2229980_8cell_rep1_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_2.fastq.gz -o $1/SRR5836473_GSM2229980_8cell_rep1_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/008/SRR5836478/SRR5836478_1.fastq.gz -o $1/SRR5836478_GSM2229985_TE_rep2_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/008/SRR5836478/SRR5836478_2.fastq.gz -o $1/SRR5836478_GSM2229985_TE_rep2_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR382/004/SRR3824224/SRR3824224_1.fastq.gz -o $1/SRR3824224_GSM2229988_ExE_rep1_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR382/004/SRR3824224/SRR3824224_2.fastq.gz -o $1/SRR3824224_GSM2229988_ExE_rep1_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR382/005/SRR3824225/SRR3824225_1.fastq.gz -o $1/SRR3824225_GSM2229989_ExE_rep2_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR382/005/SRR3824225/SRR3824225_2.fastq.gz -o $1/SRR3824225_GSM2229989_ExE_rep2_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/009/SRR5836479/SRR5836479_1.fastq.gz -o $1/SRR5836479_GSM2229987_Epiblast_rep2_WGBS_Mus_musculus_Bisulfite-Seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/009/SRR5836479/SRR5836479_2.fastq.gz -o $1/SRR5836479_GSM2229987_Epiblast_rep2_WGBS_Mus_musculus_Bisulfite-Seq_2.fastq.gz
```

–°–∫–∞—á–∞–µ–º `.fastq` —Ñ–∞–π–ª—ã –¥–ª—è –Ω—É–∂–Ω—ã—Ö –Ω–∞–º –æ–±—Ä–∞–∑—Ü–æ–≤

```bash
$ bash scripts/load_fastq.sh fastq_files
```

–ü–µ—Ä–µ–∏–º–µ–Ω—É–µ–º —Ñ–∞–π–ª—ã —á—Ç–æ–±—ã —Å –Ω–∏–º–∏ –±—ã–ª–æ –ø—Ä–æ—â–µ —Ä–∞–±–æ—Ç–∞—Ç—å. –î–ª—è —ç—Ç–æ–≥–æ —Å–¥–µ–ª–∞–µ–º –Ω–µ–±–æ–ª—å—à–æ–π —Å–∫—Ä–∏–ø—Ç –Ω–∞ `python` 

```python
import os
import glob

dir_name = 'fastq_files'
for filename in glob.glob(f'{dir_name}/*'):
    sra_code = filename.split('/')[1].split('_')[0]
    new_name = f'{dir_name}/{sra_code}.fastq.gz'
    os.rename(filename, new_name)
    print(sra_code)

```

–ó–∞–ø—É—Å—Ç–∏–º —Å–∫—Ä–∏–ø—Ç

```bash
$ python3 scripts/rename_fastq.py
```

# QC –¥–ª—è —Å—ã—Ä—ã—Ö –æ–±—Ä–∞–∑—Ü–æ–≤

–í —ç—Ç–æ—Ç —Ä–∞–∑ –±—É–¥–µ–º –∑–∞–ø—É—Å–∫–∞—Ç—å –≤—Å–µ —Ç—É–ª—ã –ø—Ä–∏ –ø–æ–º–æ—â–∏ `singularity`. –î–ª—è —ç—Ç–æ–≥–æ –Ω—É–∂–Ω–æ –∞–∫—Ç–∏–≤–∏—Ä–æ–≤–∞—Ç—å –º–æ–¥—É–ª—å `singularity` –Ω–∞ –∫–ª–∞—Å—Ç–µ—Ä–µ. –°–∫–∞—á–∏–≤–∞—Ç—å –æ–±—Ä–∞–∑—ã —Ç—É–ª–æ–≤ –±—É–¥–µ–º –ø–æ –º–µ—Ä–µ –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ—Å—Ç–∏.

<aside>
üí° –ê–Ω–∞–ª–æ–≥–∏—á–Ω–æ –º–æ–∂–Ω–æ —Ä–∞–±–æ—Ç–∞—Ç—å –∏ —Å `docker`

</aside>

–ü–æ–¥–≥—Ä—É–∑–∏–º –º–æ–¥—É–ª—å –∏ —Å–∫–∞—á–∞–µ–º –æ–±—Ä–∞–∑—ã –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã—Ö —Ç—É–ª–æ–≤. –ù–∞–∑–≤–∞–Ω–∏—è –¥–ª—è –æ–±—Ä–∞–∑–æ–≤ —è –æ–±—ã—á–Ω–æ –Ω–∞—Ö–æ–∂—É –Ω–∞ —Å–∞–π—Ç–µ [bioconda](https://bioconda.github.io/recipes/fastqc/README.html) –ø–µ—Ä–µ—Ö–æ–¥—è –ø–æ —Å—Å—ã–ª–∫–µ `quay.io/biocontainers/<toolname>` . –ù–∞–ø—Ä–∏–º–µ—Ä –≤–æ—Ç [–∑–¥–µ—Å—å](https://quay.io/repository/biocontainers/fastqc?tab=tags&tag=latest) –º–æ–∂–Ω–æ –ø–æ—Å–º–æ—Ç—Ä–µ—Ç—å –æ–±—Ä–∞–∑—ã –¥–ª—è `fastqc`

```jsx
$ module load singularity/3.6.1
$ mkdir singularityImages
$ singularity pull --dir singularityImages fastqc.sif docker://quay.io/biocontainers/fastqc:0.12.1--hdfd78af_0
$ singularity pull --dir singularityImages multiqc.sif docker://quay.io/biocontainers/multiqc:1.21--pyhdfd78af_0
```

–¢–µ–ø–µ—Ä—å —Å–¥–µ–ª–∞–µ–º —Ñ–∞–π–ª —Å –∑–∞–ø—É—Å–∫–æ–º –∞–Ω–∞–ª–∏–∑–∞ –∫–∞—á–µ—Å—Ç–≤–∞ –ø—Ä–æ—á—Ç–µ–Ω–∏–π.

```bash
#!/bin/bash
#PBS -N quality_control
#PBS -l walltime=2:30:00
#PBS -l nodes=1:ppn=8

### Change directory
cd ~/compbio/methylation

### Load singularity
module load singularity/3.6.1 

### Run QC
singularity run singularityImages/fastqc.sif fastqc -o qc -t 8 fastq_files/*.fastq.gz
singularity run singularityImages/multiqc.sif multiqc qc --outdir qc --filename multiqc.html --title "Read quality" --force
```

–ó–∞–ø—É—Å—Ç–∏–º —Å–∫—Ä–∏–ø—Ç –Ω–∞ –≤—ã—á–∏—Å–ª–∏—Ç–µ–ª—å–Ω–æ–π –Ω–æ–¥–µ –ø—Ä–∏ –ø–æ–º–æ—â–∏ `qsub`

```bash
$ qsub scripts/run_qc.sh
```

- *–†–µ–∑—É–ª—å—Ç–∞—Ç—ã*
    
    [multiqc_report.html](figures/multiqc_report.html)
    

# –¢—Ä–∏–º–º–∏–Ω–≥

–ë—É–¥–µ–º –¥–µ–ª–∞—Ç—å —Ç—Ä–∏–º–º–∏–Ω–≥ –ø—Ä–∏ –ø–æ–º–æ—â–∏ `fastp`. –°–∫–∞—á–∞–µ–º –µ–≥–æ –æ–±—Ä–∞–∑

```bash
$ singularity pull --dir singularityImages fastp.sif docker://quay.io/biocontainers/fastp:0.23.4--hadf994f_3
```

–ù–∞–º –Ω—É–∂–Ω–æ –±—É–¥–µ—Ç —Å–æ–∑–¥–∞—Ç—å –¥–≤–∞ –æ—Ç–¥–µ–ª—å–Ω—ã—Ö —Å–∫—Ä–∏–ø—Ç–∞: –æ–¥–∏–Ω –∫–æ—Ç–æ—Ä—ã–π –±—É–¥–µ—Ç –æ—Ç–ø—Ä–∞–≤–ª—è—Ç—å –∑–∞–ø—Ä–æ—Å—ã –Ω–∞ –≤—ã—á–∏—Å–ª–∏—Ç–µ–ª—å–Ω—É—é –Ω–æ–¥—É –¥–ª—è –∫–∞–∂–¥–æ–≥–æ –æ–±—Ä–∞–∑—Ü–∞, –∏ –≤—Ç–æ—Ä–æ–π, –∫–æ—Ç–æ—Ä—ã–π –±—É–¥–µ—Ç –≤—ã–ø–æ–ª–Ω—è—Ç—å —Ç—Ä–∏–º–º–∏–Ω–≥. –¢–∞–∫ –º—ã —Å–º–æ–∂–µ–º –∑–∞–¥–µ–π—Å—Ç–≤–æ–≤–∞—Ç—å –Ω–µ—Å–∫–æ–ª—å–∫–æ –Ω–æ–¥ –æ–¥–Ω–æ–≤—Ä–µ–º–µ–Ω–Ω–æ.

–¢—Ä–∏–º–º–∏–Ω–≥ –±—É–¥–µ–º –¥–µ–ª–∞—Ç—å —Ç–æ–ª—å–∫–æ –ø–æ —Å—Ä–µ–¥–Ω–µ–º—É –∫–∞—á–µ—Å—Ç–≤—É –∏ –º–∏–Ω–∏–º–∞–ª—å–Ω–æ–π –¥–ª–∏–Ω–µ –∫–∞–∫ —ç—Ç–æ –¥–µ–ª–∞–µ—Ç –ø—Ä–∏ –±–∞–∑–æ–≤—ã—Ö –Ω–∞—Å—Ç—Ä–æ–π–∫–∞—Ö `trim-galore`

```bash
#!/bin/bash
#PBS -l walltime=10:30:00
#PBS -l nodes=1:ppn=8

### Change directory
cd ~/compbio/methylation

### Load singularity
module load singularity/3.6.1 

### Trimming
singularity run singularityImages/fastp.sif fastp -i fastq_files/${SAMPLE}_1.fastq.gz -I fastq_files/${SAMPLE}_2.fastq.gz \
-o fastq_trimmed/${SAMPLE}_trimmed_1.fastq.gz -O fastq_trimmed/${SAMPLE}_trimmed_2.fastq.gz \
-h qc/${SAMPLE}_fastp.html -j qc/${SAMPLE}_fastp.json --average_qual 20 --length_required 20 -w 8
```

–¢–µ–ø–µ—Ä—å –∑–∞–ø—É—Å—Ç–∏–º –≤—ã–ø–æ–ª–Ω–µ–Ω–∏–µ —Å–∫—Ä–∏–ø—Ç–∞ –¥–ª—è –≤—Å–µ—Ö –æ–±—Ä–∞–∑—Ü–æ–≤

```bash
$ for infile in fastq_files/*1.fastq.gz
	do
	  base=$(basename -a -s _1.fastq.gz ${infile})
	  echo ${base}
	  qsub -v SAMPLE=$base -N $base scripts/trim.sh
	done
```

–û–±—ä–µ–¥–∏–Ω–∏–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç—ã `fastp` –≤ –æ–¥–∏–Ω –æ—Ç—á—ë—Ç

```bash
$ singularity run singularityImages/multiqc.sif multiqc . -f -d -dd 3 --outdir qc --filename mutiqc_report.html --force --no-data-dir
```

- –†–µ–∑—É–ª—å—Ç–∞—Ç—ã
    
    [multiqc_report.html](figures/multiqc_report.html)
    

# Bismark

## –°–∫–∞—á–∞–µ–º —Ñ–∞–π–ª—ã —Å —Ä–µ—Ñ–µ—Ä–µ–Ω—Å–æ–º

–°–æ—Å—Ç–∞–≤–∏–º —Å–∫—Ä–∏–ø—Ç –¥–ª—è —Å–∫–∞—á–∏–≤–∞–Ω–∏—è

```bash
#!/bin/bash

# —Å–∫–∞—á–∞–µ–º —Å–±–æ—Ä–∫—É 17-–æ–π —Ö—Ä–æ–º–æ—Å–æ–º—ã
wget https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.chromosome.17.fa.gz
gzip -d Mus_musculus.GRCm39.dna.chromosome.17.fa.gz
mv Mus_musculus.GRCm39.dna.chromosome.17.fa references/chr17.fa

# —Å–∫–∞—á–∞–µ–º –≥–µ–Ω–æ–º–Ω—É—é –∞–Ω–Ω–æ—Ç–∞—Ü–∏—é —Å ENSEMBL
wget https://ftp.ensembl.org/pub/release-111/gtf/mus_musculus/Mus_musculus.GRCm39.111.gtf.gz
gzip -d Mus_musculus.GRCm39.111.gtf.gz
mv Mus_musculus.GRCm39.111.gtf references/genome.gtf
```

–ó–∞–ø—É—Å—Ç–∏–º —Å–∫—Ä–∏–ø—Ç

```bash
$ bash scripts/load_reference.sh
```

## –ö–∞—Ä—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ

–°–∫–∞—á–∞–µ–º –æ–±—Ä–∞–∑ –¥–ª—è `bismark`

```bash
$ singularity pull --dir singularityImages bismark.sif docker://quay.io/biocontainers/bismark:0.24.2--hdfd78af_0
```

–ü—Ä–æ–∏–Ω–¥–µ–∫—Å–∏—Ä—É–µ–º –≥–µ–Ω–æ–º. –¢—É—Ç –Ω—É–∂–Ω–æ –±—ã—Ç—å –æ—Å—Ç–æ—Ä–æ–∂–Ω—ã–º —Å –ø–∞—Ä–∞–º–µ—Ç—Ä–æ–º `-p` —Ç.–∫. –æ–Ω –∑–∞–¥–∞—ë—Ç –≤ –¥–≤–∞ —Ä–∞–∑–∞ –±–æ–ª—å—à–µ –¥–æ—Å—Ç—É–ø–Ω—ã—Ö –ø–æ—Ç–æ–∫–æ–≤ (–ø–∞—Ä–∞–ª–ª–µ–ª—å–Ω–æ –∏–Ω–¥–µ–∫—Å–∏—Ä—É–µ—Ç—Å—è –≥–µ–Ω–æ–º –¥–ª—è –¥–≤—É—Ö –∑–∞–º–µ–Ω)

```bash
$ singularity run singularityImages/bismark.sif bismark_genome_preparation --verbose --parallel 16 references/
```

–¢–µ–ø–µ—Ä—å –Ω–∞–ø–∏—à–µ–º —Å–∫—Ä–∏–ø—Ç –¥–ª—è –∫–∞—Ä—Ç–∏—Ä–æ–≤–∞–Ω–∏—è –Ω–∞—à–∏—Ö –æ–±—Ä–∞–∑—Ü–æ–≤.

```bash
#!/bin/bash
#PBS -l walltime=10:30:00
#PBS -l nodes=1:ppn=8

### Change directory
cd ~/compbio/methylation

### Make output dir
mkdir -p bismarkOutput/${SAMPLE}

### Load singularity
module load singularity/3.6.1 

### Alignment
singularity run singularityImages/bismark.sif bismark --parallel -p 4 --gzip  references  -1 fastq_trimmed/${SAMPLE}_trimmed_1.fastq.gz  -2 fastq_trimmed/${SAMPLE}_trimmed_2.fastq.gz --output_dir bismarkOutput/${SAMPLE} &> bismarkOutput/${SAMPLE}/bismark.log 
```

<aside>
üí° –°—Ç–æ–∏—Ç –±—ã—Ç—å –æ—Å—Ç–æ—Ä–æ–∂–Ω—ã–º —Å —É–∫–∞–∑–∞–Ω–∏–µ–º`--parallel` –∏ `-p` —Ñ–ª–∞–≥–æ–≤. –°—Ç–∞–Ω–¥–∞—Ä—Ç–Ω–æ –±–∏—Å–º–∞—Ä–∫ –¥–ª—è –∫–∞–∂–¥–æ–≥–æ —Å—ç–º–ø–ª–∞ –≤—ã–ø–æ–ª–Ω—è–µ—Ç –ø–æ 2 –≤—ã—Ä–∞–≤–Ω–∏–≤–∞–Ω–∏—è (—Ç.–∫. —Å—á–∏—Ç–∞–µ—Ç, —á—Ç–æ –ø—Ä–æ—Ç–æ–∫–æ–ª –Ω–∞–ø—Ä–∞–≤–ª–µ–Ω–Ω—ã–π/stranded). –°–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ, –æ–Ω –±—É–¥–µ—Ç –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å **2 * parallel * p** –ø–æ—Ç–æ–∫–æ–≤.

`--parallel` ‚Äî —á–∏—Å–ª–æ –∏–Ω—Å—Ç–∞–Ω—Å–æ–≤ `bismark` –∫–æ—Ç–æ—Ä—ã–µ –±—É–¥—É—Ç –∑–∞–ø—É—â–µ–Ω—ã –ø–∞—Ä–∞–ª–ª–µ–ª—å–Ω–æ (–¥–µ–ª–∏—Ç           `.fastq` —Ñ–∞–π–ª –Ω–∞ —á–∞—Å—Ç–∏ –∏ –∑–∞–ø—É—Å–∫–∞–µ—Ç `bismark` –Ω–∞ –∫–∞–∂–¥–æ–π —á–∞—Å—Ç–∏ –ø–æ –æ—Ç–¥–µ–ª—å–Ω–æ—Å—Ç–∏)
`-p`  ‚Äî —á–∏—Å–ª–æ –ø–æ—Ç–æ–∫–æ–≤, –∫–æ—Ç–æ—Ä—ã–µ –∏—Å–ø–æ–ª—å–∑—É–µ—Ç bowtie –ø—Ä–∏ –≤—ã—Ä–∞–≤–Ω–∏–≤–∞–Ω–∏–∏

–¢–∞–∫ –∂–µ —Ä–∞–∑—Ä–∞–±–æ—Ç—á–∏–∫–∏ –ø—Ä–æ—Å—è—Ç –Ω–µ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –≤—ã—Å–æ–∫–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è `-p` —Ç–∞–∫ –∫–∞–∫ —ç—Ç–æ –º–æ–∂–µ—Ç –ø—Ä–∏–≤–æ–¥–∏—Ç—å –∫ –Ω–µ—Å—Ç–∞–±–∏–ª—å–Ω–æ—Å—Ç—è–º –≤ —Ä–∞–±–æ—Ç–µ.

</aside>

–¢–µ–ø–µ—Ä—å –∑–∞–ø—É—Å—Ç–∏–º –≤—ã–ø–æ–ª–Ω–µ–Ω–∏–µ —Å–∫—Ä–∏–ø—Ç–∞ –¥–ª—è –≤—Å–µ—Ö –æ–±—Ä–∞–∑—Ü–æ–≤

```bash
$ for infile in fastq_files/*1.fastq.gz
	do
	  base=$(basename -a -s _1.fastq.gz ${infile})
	  echo ${base}
	  qsub -v SAMPLE=$base -N $base scripts/bismark_align.sh
	done
```

–ü–æ—Å–º–æ—Ç—Ä–∏–º –Ω–∞ –æ–¥–∏–Ω –∏–∑ .bam —Ñ–∞–π–ª–æ–≤. –î–ª—è —ç—Ç–æ–≥–æ –≤–æ—Å–ø–æ–ª—å–∑—É–µ–º—Å—è –∫–æ–º–∞–Ω–¥–æ–π `samtools view`  (–æ–Ω —É–∂–µ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω –≤ –æ–±—Ä–∞–∑–µ `bismark` —Ç–∞–∫ —á—Ç–æ –º–æ–∂–Ω–æ –¥–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω–æ –Ω–µ —Å–∫–∞—á–∏–≤–∞—Ç—å)

```bash
$ singularity run singularityImages/bismark.sif samtools view bismarkOutput/SRR3824224/SRR3824224_trimmed_1_bismark_bt2_pe.bam | head -n 5

SRR3824225.1_1/1        83      17      76346865        0       140M    =       76346848   -157     AACAAATAAAATAAAAACAAAATAAACCTTTAATAACCTTAAAAAAAATAACTACCCCTCTTTTATTTTAAAAAACCAATTCTTTAAAATACAATAAACACATTCAACTATTCCTTATCCCATAAAATTATAAAAAATTC        ????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????    NM:i:39 MD:Z:0G8G0G1G2G0G3G0G1G0G0G5G0G1G5G0G0G0G0G1T3G2G10G9G12G0G1G1G2G1G3G14G6G0G0G0G2G3G0G5     XM:Z:h........hh.h..hh...hh.hhh.....hh.h.....hhhhh.....h..x..........h.........h............hh.h.z..h.h...z..............h......hhhh..h...hh.....   XR:Z:CT     XG:Z:GA
SRR3824225.1_1/1        163     17      76346848        0       140M    =       76346865   157      GAGGACAAAAAATTTAAAACAAATAAAATAAAAACAAAATAAACCTTTAATAACCTTAAAAAAAATAACTACCCCTCTTTTATTTTAAAAAACCAATTCTTTAAAATACAATAAACACATTCAACTATTCCTTATCCCAT        ????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????    NM:i:37 MD:Z:1T1T0G3G7G0G8G0G1G2G0G3G0G1G0G0G5G0G1G5G0G0G0G0G1T3G2G10G9G12G0G1G1G2G1G3G14G6 XM:Z:X.H.h...h.......hh........hh.h..hh...hh.hhh.....hh.h.....hhhhh.....h..x..........h.........h............hh.h.z..h.h...z..............h......   XR:Z:GA XG:Z:GA
SRR3824225.26_26/1      99      17      41386106        0       140M    =       41386123   157      AAATTTTAAATTAGATTGAAGGTTATGTGTTGGAATGTTTTGGGAATGAAGATAGTAGTTGTTATTTGTGTGTTTTAGGGGGGTGGGGATGTATATTTATTGGTTCGGTTGGAGAATGTTGCTTAATTTGTTTGAAAAAT        ????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????    NM:i:30 MD:Z:6C4C3C0C5C0C5C8C0C7A7C5C0C1C0C6C1C10C5C4C0C0G0C4C3C12A4C0C2C8C0        XM:Z:......h....x...xz.....hh.....x........hx...............x.....hh.hx......h.h................h....xz.x....xZ..x.................hx..h........h   XR:Z:CT XG:Z:CT
SRR3824225.26_26/1      147     17      41386123        0       140M    =       41386106   -157     GAAGGTTATGTGTTGGAATGTTTTGGGAATGAAGATAGTAGTTGTTATTTGTGTGTTTTAGGGGGGTGGGGATGTATATTTATTGGTTCGGTTGGAGAATGTTGATTAATTTGTTTGAAAAATAAAATTTAGATAGTGAG        ????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????    NM:i:27 MD:Z:5C0C5C8C0C7A7C5C0C1C0C6C1C10C5C4C0C0G0C4C3C17C0C2C8C5C4C6      XM:Z:.....hh.....x........hx...............x.....hh.hx......h.h................h....xz.x....xZ..x.................hx..h........h.....h....x......   XR:Z:GA XG:Z:CT
SRR3824225.28_28/1      99      17      41386053        0       140M    =       41386120   207      ATTAAAATAAAAATTAATTTTTTATTTGTTTTTTTAGTTTTTTTGGTAAATTTAAATTTTAAATTAGATTGAAGGTTACGTGTTGGAATGTTTTGGGAATGAAGATAGTAGTTGTTATTTGTGTGTTTTACGGGGGTGGG        ????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????    NM:i:35 MD:Z:2C10C0C3C1C4C2C1C2C3C0C0C1C8C0C7C4C3C0C5C0C1T3C8C0C7A7C5C0C1C0C6C1C2G7C1       XM:Z:..h..........hh...h.h....x..h.h..h...hhh.h........hh.......h....x...xz.....hh.....x........hx...............x.....hh.hx......h.h............   XR:Z:CT XG:Z:CT
```

–í–∏–¥–Ω–æ, —á—Ç–æ –Ω–µ—Å–º–æ—Ç—Ä—è –Ω–∞ —Å—Ç—Ä–∞–Ω–Ω—ã–µ –Ω–∞–∑–≤–∞–Ω–∏—è —Ñ–∞–π–ª–æ–≤ –∏ —Ä–∏–¥–æ–≤ .bam —Ñ–∞–π–ª—ã —Å–æ–¥–µ—Ä–∂–∞—Ç –∏ –ø—Ä—è–º—ã–µ –∏ –æ–±—Ä–∞—Ç–Ω—ã–µ –ø—Ä–æ—á—Ç–µ–Ω–∏—è. –û–ø–∏—Å–∞–Ω–∏–µ —Ñ–ª–∞–≥–æ–≤ —Ä–∏–¥–æ–≤ (–≤—Ç–æ—Ä–∞—è –∫–æ–ª–æ–Ω–∫–∞) –º–æ–∂–Ω–æ –ø–æ—Å–º–æ—Ç—Ä–µ—Ç—å –≤–æ—Ç –∑–¥–µ—Å—å

[Explain SAM Flags](https://broadinstitute.github.io/picard/explain-flags.html)

## –î–µ–¥—É–ø–ª–∏–∫–∞—Ü–∏—è

–¢–µ–ø–µ—Ä—å —Å–¥–µ–ª–∞–µ–º –¥–µ–¥—É–ø–ª–∏–∫–∞—Ü–∏—é —á—Ç–æ–±—ã —É–±—Ä–∞—Ç—å –≤–æ–∑–º–æ–∂–Ω—ã–µ –∞—Ä—Ç–µ—Ñ–∞–∫—Ç—ã PCR

```bash
#!/bin/bash

# Load singularity
module load singularity/3.6.1

# Dedupliaction step
for infile in bismarkOutput/*/*.bam
do
    base=$(basename -a -s _trimmed_1_bismark_bt2_pe.bam ${infile})
    echo ${base}
    singularity run singularityImages/bismark.sif deduplicate_bismark -p --output_dir bismarkOutput/${base} $infile
done
```

–ó–∞–ø—É—Å—Ç–∏–º –≤—ã–ø–æ–ª–Ω–µ–Ω–∏–µ –¥–µ–¥—É–ø–ª–∏–∫–∞—Ü–∏–∏

```bash
$ bash scripts/deduplicate.sh
```

## –ü–æ–ª—É—á–µ–Ω–∏–µ –ø—Ä–æ—Ñ–∏–ª—è –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è –î–ù–ö

–î–ª—è –∏–∑–≤–ª–µ—á–µ–Ω–∏—è –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏–∏ –æ –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏–∏ —Ü–∏—Ç–æ–∑–∏–Ω–æ–≤ –ø–æ –≤—Å–µ–º—É –≥–µ–Ω–æ–º—É –∏–∑ –ø–æ–ª—É—á–µ–Ω–Ω—ã—Ö –≤—ã—Ä–∞–≤–Ω–∏–≤–∞–Ω–∏–π –∑–∞–ø—É—Å–∫–∞–µ—Ç—Å—è –ø—Ä–æ–≥—Ä–∞–º–º–∞ `bismark_methylation_extractor`

```bash
#!/bin/bash

# Load singularity
module load singularity/3.6.1

# Methylatioin calling step
for infile in bismarkOutput/*/*deduplicated.bam
do
    base=$(basename -a -s _trimmed_1_bismark_bt2_pe.deduplicated.bam ${infile})
    echo ${base}
    singularity run singularityImages/bismark.sif bismark_methylation_extractor -p --output_dir bismarkOutput/${base} \
    --no_overlap --gzip --bedGraph --cytosine_report $infile
done
```

–°–æ–∑–¥–∞–¥–∏–º –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é —Å –∞—É—Ç–ø—É—Ç–æ–º –∏ –∑–∞–ø—É—Å—Ç–∏–º —Å–∫—Ä–∏–ø—Ç

```bash
$ mkdir methylationOutput
$ bash scripts/methcall.sh
```

## –°–æ–∑–¥–∞–Ω–∏–µ –æ—Ç—á—ë—Ç–∞

–ó–∞–ø—É—Ç–∏–º `multiqc` —á—Ç–æ–±—ã —Å–æ–±—Ä–∞—Ç—å —Ä–µ–∑—É–ª—å—Ç–∞—Ç—ã –≤–º–µ—Å—Ç–µ

```bash
$ singularity run singularityImages/multiqc.sif multiqc . -f -d -dd 3 --outdir qc --filename bismark_report.html --force --no-data-dir
```

- *–†–µ–∑—É–ª—å—Ç–∞—Ç*
    
    [bismark_report.html](figures/bismark_report.html)
    

## –ò–∑–≤–ª–µ—á–µ–º –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏–∏ –≤ CpG –∫–æ–Ω—Ç–µ–∫—Å—Ç–µ

## –í–∏–∑—É–∞–ª–∏–∑–∞—Ü–∏—è –≤ –≥–µ–Ω–æ–º–Ω–æ–º –±—Ä–∞—É–∑–µ—Ä–µ

–°–∫–∞—á–∞–µ–º –æ–±—Ä–∞–∑ –¥–ª—è `bedtools`

```bash
$ singularity pull --dir singularityImages bedtools.sif docker://quay.io/biocontainers/bedtools:2.31.1--hf5e1c6e_1
```

–¢–µ–ø–µ—Ä—å —Å–æ–∑–¥–∞–¥–∏–º `.bedGraph` —Ñ–∞–π–ª—ã, –∫–æ—Ç–æ—Ä—ã–µ —Å–æ–¥–µ—Ä–∂–∞—Ç –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ –ø–æ–∫—Ä—ã—Ç–∏–∏. –î–ª—è —ç—Ç–æ–≥–æ –æ—Ç—Å–æ—Ä—Ç–∏—Ä—É–µ–º `.bam` —Ñ–∞–π–ª—ã –∏ –≤–æ—Å–ø–æ–ª—å–∑—É–µ–º—Å—è —Ç—É–ª–æ–º `genomecov`

```bash
#!/bin/bash

for file in bismarkOutput/*/*.deduplicated.bam
do
    base=$(basename -a -s _trimmed_1_bismark_bt2_pe.deduplicated.bam ${file})
    singularity run singularityImages/samtools.sif samtools sort -@16 $file -o bismarkOutput/${base}/${base}.deduplicated.sorted.bam
    singularity run singularityImages/bedtools.sif bedtools genomecov \
    -ibam bismarkOutput/${base}/${base}.deduplicated.sorted.bam -bg > bismarkOutput/${base}/${base}.coverage.bedGraph
done
```

–ó–∞–ø—É—Å—Ç–∏–º —Å–∫—Ä–∏–ø—Ç

```bash
$ bash scripts/bedGraph_coverage.sh
```

–¢–µ–ø–µ—Ä—å —Å–∫–æ–ø–∏—Ä—É–µ–º .bedGraph —Ñ–∞–π–ª—ã –≤ –æ—Ç –¥–µ–ª—å–Ω—É—é –ø–∞–ø–∫—É, —Ä–∞–∑–∞—Ä—Ö–∏–≤–∏—Ä—É–µ–º (—Ç–µ –∫–æ—Ç–æ—Ä—ã–µ –∞—Ä—Ö–∏–≤–∏—Ä–æ–≤–∞–Ω—ã) –∏ —Å–∫–∞—á–∞–µ–º –Ω–∞ –∫–æ–º–ø—å—é—Ç–µ—Ä.

  

```bash
$ mkdir bismarkBedGraph # —Å–æ–∑–¥–∞–Ω–∏–µ –Ω–æ–≤–æ–π –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏–∏
$ cp bismarkOutput/*/*.bedGraph* bismarkBedGraph/ # –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏–µ .bedGraph –§–∞–π–ª–æ–≤ –±–∏—Å–º–∞—Ä–∫–∞
$ for file in bismarkBedGraph/*.gz; do gzip -d $file; done # —Ä–∞–∑–∞—Ä—Ö–∏–≤–∏—Ä—É–µ–º —Ñ–∞–π–ª—ã
$ tar -cvf bismarkBedGraph.tar bismarkBedGraph/*
```

–¢–µ–ø–µ—Ä—å –∑–∞–≥—Ä—É–∑–∏–º –≤—Å–µ —Ñ–∞–π–ª—ã (–≥–µ–Ω–æ–º –º—ã—à–∫–∏ + `.bedGraph`) –Ω–∞ —Å–∞–π—Ç–µ –ø–µ—Ä—Å–µ—Ñ–æ–Ω—ã

![Untitled](figures/Untitled.png)

# –ê–Ω–∞–ª–∏–∑ –¥–∏—Ñ—Ñ–µ—Ä–µ–Ω—Ü–∏–∞–ª—å–Ω–æ–≥–æ –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è

## –ü–æ–¥–≥–æ—Ç–æ–≤–∫–∞

–ë—É–¥–µ–º —Å—Ä–∞–≤–Ω–∏–≤–∞—Ç—å `8cell vs Epiblast` 

| 8cell | Epiblast |
| --- | --- |
| SRR5836474 | SRR3824222 |
| SRR5836473 | SRR5836479 |

–£ –Ω–∞—Å –¥–æ—Å—Ç–∞—Ç–æ—á–Ω–æ –≤—ã—Å–æ–∫–∞—è –≥–ª—É–±–∏–Ω–∞ —Å–µ–∫–≤–µ–Ω–∏—Ä–æ–≤–∞–Ω–∏—è –¥–ª—è –Ω–∞—à–∏—Ö –æ–±—Ä–∞–∑—Ü–æ–≤. –ü–æ—ç—Ç–æ–º—É –º–æ–∂–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —Ç—É–ª—ã, –∫–æ—Ç–æ—Ä—ã–µ –Ω–µ –¥–µ–ª–∞—é—Ç —Å–≥–ª–∞–∂–∏–≤–∞–Ω–∏–µ. –ü–æ–ø—Ä–æ–±—É–µ–º –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å `methylKit` —Ç.–∫. –∫ –Ω–µ–º—É –µ—Å—Ç—å –Ω–µ–ø–ª–æ—Ö–∞—è –¥–æ–∫—É–º–µ–Ω—Ç–∞—Ü–∏—è

- –ü–æ–ª–µ–∑–Ω—ã–µ –º–∞—Ç–µ—Ä–∏–∞–ª—ã
    
    [DNA Methylation Analysis using methylKit](https://rosericazondekon.github.io/posts/methylkit-dna-methylation-analysis/)
    
    [10.5 Extracting interesting regions: Differential methylation and segmentation | Computational Genomics with R](https://compgenomr.github.io/book/extracting-interesting-regions-differential-methylation-and-segmentation.html)
    
    [methylKit.knit](https://www.bioconductor.org/packages/release/bioc/vignettes/methylKit/inst/doc/methylKit.html#57_methSeg():_segmentation_of_methylation_or_differential_methylation_profiles)
    

## –ê–Ω–∞–ª–∏–∑

[https://github.com/Claptar/methylation_mipt/blob/master/methylkit.Rmd](https://github.com/Claptar/methylation_mipt/blob/master/methylkit.Rmd)

–£—Å—Ç–∞–Ω–æ–≤–∏–º –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –±–∏–±–ª–∏–æ—Ç–µ–∫–∏

```r
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("methylKit")
BiocManager::install("biomaRt")
```

–ü–æ–¥–≥—Ä—É–∑–∏–º –±–∏–±–ª–∏–æ—Ç–µ–∫–∏

```r
library(methylKit)
library(biomaRt)
```

–ü–æ–¥–≥—Ä—É–∑–∏–º —Ñ–∞–π–ª—ã

```r
file.list = list(
'bismarkMethylation/SRR5836473_trimmed_1_bismark_bt2_pe.deduplicated.bismark.cov',
'bismarkMethylation/SRR5836474_trimmed_1_bismark_bt2_pe.deduplicated.bismark.cov',
'bismarkMethylation/SRR3824222_trimmed_1_bismark_bt2_pe.deduplicated.bismark.cov',
'bismarkMethylation/SRR5836479_trimmed_1_bismark_bt2_pe.deduplicated.bismark.cov'
)
```

```r
# Read the files into a methylKit object of type methylRawList: myobj
myobj = methRead(file.list,
                 pipeline="bismarkCoverage",
                 sample.id=list("8cell.1","8cell.2","Epiblast.1","Epiblast.2"),
                 assembly="mm10",
                 treatment=c(0, 0, 1, 1),
                 context="CpG",
                 )
```

–ü–æ—Å–º–æ—Ç—Ä–∏–º –Ω–∞ —Ä–∞—Å–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω—ã—Ö —Å–∞–π—Ç–æ–≤

```r
getMethylationStats(myobj[[1]],plot=TRUE,both.strands=FALSE)
getMethylationStats(myobj[[2]],plot=TRUE,both.strands=FALSE)
getMethylationStats(myobj[[3]],plot=TRUE,both.strands=FALSE)
getMethylationStats(myobj[[4]],plot=TRUE,both.strands=FALSE)
```

- –†–µ–∑—É–ª—å—Ç–∞—Ç
    
    ![00001a.png](figures/00001a.png)
    
    ![00001c.png](figures/00001c.png)
    
    ![00001e.png](figures/00001e.png)
    
    ![000018.png](figures/000018.png)
    

–§–∏–ª—å—Ç—Ä–∞—Ü–∏—è –∏ –æ–±—ä–µ–¥–∏–Ω–µ–Ω–∏–µ —Ñ–∞–π–ª–æ–≤

```r
# Filter out CpG sites with too few or too many reads
filtered.myobj = filterByCoverage(myobj, lo.count=10, lo.perc=NULL, hi.count=NULL, hi.perc=99.9)

# Generate methylation matrix by combining all samples. By default it includes only those CpGs which are covered by reads in each sample.
meth = unite(filtered.myobj) 
```

–ü–æ—Å–º–æ—Ç—Ä–∏–º –Ω–∞ –∫–æ—Ä—Ä–µ–ª—è—Ü–∏—é –æ–±—Ä–∞–∑—Ü–æ–≤

```r
getCorrelation(meth,plot=TRUE)
```

- –†–µ–∑—É–ª—å—Ç–∞—Ç
    
    ![Untitled](figures/Untitled%201.png)
    

–°–¥–µ–ª–∞–µ–º PCA

```r
pc = PCASamples(meth, obj.return=TRUE, adj.lim=c(1,1))
```

- –†–µ–∑—É–ª—å—Ç–∞—Ç
    
    ![Untitled](figures/Untitled%202.png)
    

–ê–Ω–∞–ª–∏–∑ –¥–∏—Ñ—Ñ–µ—Ä–µ–Ω—Ü–∏–∞–ª—å–Ω–æ–≥–æ –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω–∏—è –±—É–¥–µ–º –¥–µ–ª–∞—Ç—å –ø—Ä–∏ –ø–æ–º–æ—â–∏ —Ö–∏-–∫–≤–∞–¥—Ä–∞—Ç–∞

```r
# Run differential analysis 
myDiff = calculateDiffMeth(meth, test = "Chisq", overdispersion ='MN')
```

–û—Å—Ç–∞–≤–∏–º —Ç–æ–ª—å–∫–æ –∑–Ω–∞—á–∏–º—ã–µ —Å–∞–π—Ç—ã

```r
# The "difference" parameter sets the percentage threshold for differences in methylation levels
myDiff25 = getMethylDiff(myDiff, difference=25, qvalue=0.001)
myDiff25
```

|  | chr<int> | start<int> | end<int> | strand<chr> | pvalue<dbl> | qvalue<dbl> | meth.diff<dbl> |
| --- | --- | --- | --- | --- | --- | --- | --- |
| 3 | 17 | 3119156 | 3119156 | * | 1.214907e-04 | 2.114906e-04 | 29.26235 |
| 5 | 17 | 3119164 | 3119164 | * | 5.827433e-07 | 2.048513e-06 | 36.32017 |
| 6 | 17 | 3119191 | 3119191 | * | 1.008455e-04 | 1.801668e-04 | 38.06514 |
| 12 | 17 | 3127785 | 3127785 | * | 5.231073e-10 | 3.844921e-09 | 39.80856 |
| 13 | 17 | 3127786 | 3127786 | * | 1.862772e-04 | 3.056103e-04 | 28.55392 |

–°–¥–µ–ª–∞–µ–º –∞–Ω–Ω–æ—Ç–∞—Ü–∏—é –¥–∏—Ñ—Ñ–µ—Ä–µ–Ω—Ü–∏–∞–ª—å–Ω–æ –º–µ—Ç–∏–ª–∏—Ä–æ–≤–∞–Ω—ã—Ö —Å–∞–π—Ç–æ–≤. –î–ª—è —ç—Ç–æ–≥–æ —Å–∫–∞—á–∞–µ–º –∞–Ω–Ω–æ—Ç–∞—Ü–∏—é Ensembl

```r
# Download genomic annotation for particular version of genome assembly (mm10)
# Might take a while, depends on cluster traffic load

ensembl111 <- useEnsembl(biomart = 'genes', dataset = 'mmusculus_gene_ensembl', version=111)

# Get genes annotation - TSSs and chromosomes
g <- getBM(attributes=c('ensembl_gene_id','chromosome_name',"transcription_start_site"), mart = ensembl111)

# Annotation contains several TSSs for some genes, but we want to keep only one for convenience
# This will not affect the results, because each CpG will be assigned to the closest TSS
g=g[!duplicated(g[, c("ensembl_gene_id")], fromLast=F),]
```

–°–æ–∑–¥–∞–¥–∏–º GRanges –æ–±—ä–µ–∫—Ç

```r
# Make a GRanges object from the genes table
genes_all = makeGRangesFromDataFrame(g,
                                     keep.extra.columns=T,
                                     ignore.strand=T,
                                     seqinfo=NULL,
                                     seqnames.field=c("chromosome_name"),
                                     start.field="transcription_start_site",
                                     end.field=c("transcription_start_site"))
```

```r
# Make a GRanges object from the table with strong differentially methylated CpGs
mydiff25_ranges = makeGRangesFromDataFrame(myDiff25, keep.extra.columns=F)
mydiff25_ranges
>>>
GRanges object with 6490 ranges and 0 metadata columns:
        seqnames    ranges strand
           <Rle> <IRanges>  <Rle>
      3       17   3119156      *
      5       17   3119164      *
      6       17   3119191      *
     12       17   3127785      *
     13       17   3127786      *
    ...      ...       ...    ...
  26410       17  95149391      *
  26411       17  95149489      *
  26412       17  95149490      *
  26413       17  95149502      *
  26414       17  95149503      *
```

–î–ª—è –∫–∞–∂–¥–æ–≥–æ —Å–∞–π—Ç–∞ –Ω–∞–π–¥—ë–º –±–ª–∏–∂–∞–π—à–∏–µ —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ç—ã

```r
# Find the nearest TSS for each CpG
x = nearest(mydiff25_ranges, genes_all)

myDiff25['nearest_tss'] = g[x,'transcription_start_site']
myDiff25['nearest_gene'] = g[x,'ensembl_gene_id']

myDiff25_x1 = myDiff25[myDiff25$meth.diff>0,]
myDiff25_x2 = myDiff25[myDiff25$meth.diff<0,]

myDiff25
```

|  | chr | start | end | strand | pvalue | qvalue | meth.diff | nearest_tss | nearest_gene |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 3 | 17 | 3119156 | 3119156 | * | 0.0001214907421 | 0.0002114906346 | 29.26234505 | 3108194 | ENSMUSG00000088244 |
| 5 | 17 | 3119164 | 3119164 | * | 5.83E-07 | 2.05E-06 | 36.32016632 | 3108194 | ENSMUSG00000088244 |
| 6 | 17 | 3119191 | 3119191 | * | 0.0001008454984 | 0.0001801667755 | 38.06514341 | 3108194 | ENSMUSG00000088244 |
| 12 | 17 | 3127785 | 3127785 | * | 5.23E-10 | 3.84E-09 | 39.80855856 | 3134315 | ENSMUSG00000121378 |
| 13 | 17 | 3127786 | 3127786 | * | 0.0001862772295 | 0.0003056103415 | 28.55392157 | 3134315 | ENSMUSG00000121378 |

–°–æ—Ö—Ä–∞–Ω–∏–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç—ã –≤ –æ—Ç–¥–µ–ª—å–Ω—ã–π —Ñ–∞–π–ª

```r
write.csv(myDiff25, "results.csv")
```

–ü–æ—Å–º–æ—Ç—Ä–∏–º –Ω–∞ –≥–µ–Ω—ã –¥–ª—è –∫–æ—Ç–æ—Ä—ã—Ö –µ—Å—Ç—å 4 –∏–ª–∏ –±–æ–ª—å—à–µ CpG

```r
# Find genes with 4 or more CpGs assigned.
gene_diffMeth_x1 = table(myDiff25_x1$nearest_gene)
gene_diffMeth_x1 = names(gene_diffMeth_x1[gene_diffMeth_x1>=4])
head(gene_diffMeth_x1, 5)
>>>
"ENSMUSG00000000127" "ENSMUSG00000002017" "ENSMUSG00000002617" "ENSMUSG00000003279" "ENSMUSG00000003929"
```

```r
# Find genes with 4 or more CpGs assigned.
gene_diffMeth_x2 = table(myDiff25_x2$nearest_gene)
gene_diffMeth_x2 = names(gene_diffMeth_x2[gene_diffMeth_x2>=4])
gene_diffMeth_x2
>>>
"ENSMUSG00000117496"
```

```r
sessionInfo()
>>>
R version 4.3.3 (2024-02-29 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 11 x64 (build 22631)

Matrix products: default

locale:
[1] LC_COLLATE=Russian_Belarus.utf8  LC_CTYPE=Russian_Belarus.utf8    LC_MONETARY=Russian_Belarus.utf8
[4] LC_NUMERIC=C                     LC_TIME=Russian_Belarus.utf8    

time zone: Europe/Moscow
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
[1] BiocManager_1.30.22  biomaRt_2.58.2       methylKit_1.28.0     GenomicRanges_1.54.1 GenomeInfoDb_1.38.8 
[6] IRanges_2.36.0       S4Vectors_0.40.2     BiocGenerics_0.48.1 

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3          jsonlite_1.8.8              magrittr_2.0.3              farver_2.1.1               
  [5] fs_1.6.4                    BiocIO_1.12.0               zlibbioc_1.48.2             vctrs_0.6.5                
  [9] memoise_2.0.1               Rsamtools_2.18.0            RCurl_1.98-1.14             ggtree_3.10.1              
 [13] progress_1.2.3              S4Arrays_1.2.1              curl_5.2.1                  SparseArray_1.2.4          
 [17] gridGraphics_0.5-1          KernSmooth_2.23-22          plyr_1.8.9                  cachem_1.0.8               
 [21] GenomicAlignments_1.38.2    igraph_2.0.3                lifecycle_1.0.4             pkgconfig_2.0.3            
 [25] Matrix_1.6-5                R6_2.5.1                    fastmap_1.1.1               gson_0.1.0                 
 [29] MatrixGenerics_1.14.0       GenomeInfoDbData_1.2.11     digest_0.6.35               numDeriv_2016.8-1.1        
 [33] aplot_0.2.2                 enrichplot_1.22.0           colorspace_2.1-0            patchwork_1.2.0            
 [37] AnnotationDbi_1.64.1        RSQLite_2.3.6               filelock_1.0.3              fansi_1.0.6                
 [41] mgcv_1.9-1                  abind_1.4-5                 httr_1.4.7                  polyclip_1.10-6            
 [45] compiler_4.3.3              bit64_4.0.5                 withr_3.0.0                 BiocParallel_1.36.0        
 [49] viridis_0.6.5               DBI_1.2.2                   ggforce_0.4.2               R.utils_2.12.3             
 [53] MASS_7.3-60.0.1             rappdirs_0.3.3              DelayedArray_0.28.0         rjson_0.2.21               
 [57] HDO.db_0.99.1               gtools_3.9.5                tools_4.3.3                 ape_5.8                    
 [61] scatterpie_0.2.2            R.oo_1.26.0                 glue_1.7.0                  restfulr_0.0.15            
 [65] nlme_3.1-164                GOSemSim_2.28.1             grid_4.3.3                  shadowtext_0.1.3           
 [69] reshape2_1.4.4              fgsea_1.28.0                generics_0.1.3              gtable_0.3.5               
 [73] R.methodsS3_1.8.2           tidyr_1.3.1                 hms_1.1.3                   data.table_1.15.4          
 [77] xml2_1.3.6                  tidygraph_1.3.1             utf8_1.2.4                  XVector_0.42.0             
 [81] ggrepel_0.9.5               pillar_1.9.0                stringr_1.5.1               yulab.utils_0.1.4          
 [85] fastseg_1.48.0              emdbook_1.3.13              limma_3.58.1                splines_4.3.3              
 [89] dplyr_1.1.4                 tweenr_2.0.3                BiocFileCache_2.10.2        treeio_1.26.0              
 [93] lattice_0.22-5              renv_1.0.7                  rtracklayer_1.62.0          bit_4.0.5                  
 [97] tidyselect_1.2.1            GO.db_3.18.0                Biostrings_2.70.3           knitr_1.46                 
[101] gridExtra_2.3               SummarizedExperiment_1.32.0 xfun_0.43                   graphlayouts_1.1.1         
[105] Biobase_2.62.0              statmod_1.5.0               matrixStats_1.3.0           stringi_1.8.3              
[109] lazyeval_0.2.2              ggfun_0.1.4                 yaml_2.3.8                  codetools_0.2-19           
[113] bbmle_1.0.25.1              ggraph_2.2.1                tibble_3.2.1                qvalue_2.34.0              
[117] ggplotify_0.1.2             cli_3.6.2                   munsell_0.5.1               Rcpp_1.0.12                
[121] dbplyr_2.5.0                coda_0.19-4.1               png_0.1-8                   bdsmatrix_1.3-7            
[125] XML_3.99-0.16.1             parallel_4.3.3              ggplot2_3.5.1               blob_1.2.4                 
[129] prettyunits_1.2.0           clusterProfiler_4.10.1      mclust_6.1                  DOSE_3.28.2                
[133] bitops_1.0-7                viridisLite_0.4.2           mvtnorm_1.2-4               tidytree_0.4.6             
[137] scales_1.3.0                purrr_1.0.2                 crayon_1.5.2                rlang_1.1.3                
[141] cowplot_1.1.3               fastmatch_1.1-4             KEGGREST_1.42.0 
```
